File: framework/test2/validate/list2d/List2d.scala
import soda.scala.unittest._

import scala.reflect.runtime.universe.typeOf

import soda.scala.leetcode._

import collection.mutable
import collection.mutable.{ArrayBuffer, ArrayDeque, PriorityQueue}
import util.control.Breaks.{break, breakable}

object Solution {
  def groupByLength(strs: Array[String]): Array[Array[String]] = {
    val group = mutable.Map[Int, ArrayBuffer[String]]()
    for (s <- scala.util.Random.shuffle(strs.toList)) {
      group.getOrElseUpdate(s.length, ArrayBuffer[String]()) += s
    }
    scala.util.Random.shuffle(group.keys.toList).map(group(_).toArray).toArray
  }
}

class List2d {
  def get(): () => Unit = {
    () => {
      val work = GenericTestWork.create1(Solution.groupByLength)
      // val work = GenericTestWork.forStruct(typeOf[STRUCT])
      // val work = TestWork.forObject(typeOf[Solution], "METHOD")
      // val work = TestWork.forInstance(typeOf[Solution], "METHOD")
      // val work = TestWork.createN((...)=>R)
      // val work = TestWork.forStruct(typeOf[STRUCT])
      work.setValidator(Validators.forArray2d[String](false, false))
      work.compareSerial = true
      work.run()
    }
  }
}

object List2d extends App {
  new List2d().get().apply()
}

File: framework/test2/validate/list2d/List2d.cs
using Soda.Unittest;
using Soda.Leetcode;

public class Solution {
    public string[][] GroupByLength(string[] strs)
    {
        var rand = new Random();
        var strList = new List<string>(strs).OrderBy(x => rand.Next()).ToList();
        var group = new Dictionary<int, IList<string>>();
        foreach (var s in strList) {
            IList<string> L = null;
            if (!group.TryGetValue(s.Length, out L)) {
                group[s.Length] = L = new List<string>();
            }
            L.Add(s);
        }
        var keys = new List<int>(group.Keys);
        keys = keys.OrderBy(x => rand.Next()).ToList();
        var res = new string[group.Count][];
        for (int i = 0; i < res.Length; ++i) {
            res[i] = group[keys[i]].ToArray();
        }
        return res;
    }
}

public class List2d
{
    public static void Main(string[] args)
    {
        var work = WorkFactory.Create(Utils.Fn(new Solution().GroupByLength));
        // var work = WorkFactory.ForStruct<STRUCT>();
        work.SetValidator(Validators.ForArray2d<string>(false, false));
        work.CompareSerial = true;
        Console.WriteLine(work.Run(Utils.ReadStdin()));
    }
}
